test/jdk/javax/swing/JFileChooser/FileSystemView/SystemIconTest.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8182043
+ * @requires os.family == "windows"
+ * @summary Access to Windows Large Icons
+ * sun.awt.shell.ShellFolder
+ * @run main SystemIconTest
+ */
+import javax.swing.ImageIcon;
+import javax.swing.filechooser.FileSystemView;
+import java.awt.image.MultiResolutionImage;
+import java.io.File;
+
+public class SystemIconTest {
+    static final FileSystemView fsv = FileSystemView.getFileSystemView();
+
+    public static void main(String[] args) {
+        testSystemIcon();
+        System.out.println("ok");
+    }
+
+    static void testSystemIcon() {
+        String windir = System.getenv("windir");
+        testSystemIcon(new File(windir));
+        testSystemIcon(new File(windir + "/explorer.exe"));
+        return;
+    }
+
+    static void testSystemIcon(File file) {
+        int[] sizes = new int[] {16, 32, 48, 64, 128};
+        for (int size : sizes) {
+            ImageIcon icon = (ImageIcon) fsv.getSystemIcon(file, size);
+
+            //Enable below to see the icon
+            //JLabel label = new JLabel(icon);
+            //JOptionPane.showMessageDialog(null, label);
+
+            if (icon == null) {
+                throw new RuntimeException("icon is null!!!");
+            }
+
+            if (icon.getIconWidth() != size) {
+                throw new RuntimeException("Wrong icon size " +
+                        icon.getIconWidth() + " when requested " + size);
+            }
+
+            if (icon.getImage() instanceof MultiResolutionImage) {
+                MultiResolutionImage mri = (MultiResolutionImage) icon.getImage();
+                if (mri.getResolutionVariant(size, size) == null) {
```

Comment:
> This is to describe one of my questions above, is this instanceof+cast cannot be improved? Why we cannot always wrap the data in the MRI and if we have only one icon return the MRI with one resolution?

No because in specification we say that we do not guarantee the the icon returned will be MRI. We know how it works on Windows so we test it this way.

Reply:
Here I am not talking about specification, I am talking about the usage of it, is the instanceof+cast is helpful? It does not look good, why we cannot always return MRI?

Reply:
> Here I am not talking about specification, I am talking about the usage of it, is the instanceof+cast is helpful?

It is necessary until all the consequences of always returning of MRI is extensively tested on all the supported platforms and it does not cause any serious regression or visual degradations. So for now i would not enforce MRI usage in this method which means use base image class instead and do the cast later.

Reply:
I did not get how you will be able to force use MRI later since this method is implemented as a return icon. So this choice should be made before integration.

Reply:
> I did not get how you will be able to force use MRI later since this method is implemented as a return icon. So this choice should be made before integration.

I am not going to force MRI usage - i am assuming that there might be an implementation that does not do it hence i am going to go to the lowest common denominator - Icon.

Reply:
But in the code above you do not use the icon, you use an ImageIcon. So to get the resolution variant the user need to do something like:
```
    Icon icon = fsv.getSystemIcon(file, size);
    if (icon instanceof ImageIcon) {
        ImageIcon imageIcon = (ImageIcon) icon;
        if (icon.getImage() instanceof MultiResolutionImage) {
            MultiResolutionImage mri = (MultiResolutionImage) icon.getImage();
            .... = mri.getResolutionVariant(size, size);
        }
    }
```
Probably some other variants were discussed already and we cannot do something better?